### PR TITLE
Client side encryption v2

### PR DESCRIPTION
**Pull request by rullzer**
Work in progress by @tcanabrava

### CODE REVIEW

File: src/libsync/clientsideencryption.cpp [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+            handleErrors();
+        }
+
+        /* Initialise the encryption operation. */
+        if(1 != EVP_EncryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL)) {
+            handleErrors();
+        }
+
+        /* Set IV length if default 12 bytes (96 bits) is not appropriate */
+        if(1 != EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, 16, NULL)) {
```
Comment (leonklingele) [inline src/libsync/clientsideencryption.cpp]: Why use a non-standard nonce size?

Comment (tcanabrava) [reply src/libsync/clientsideencryption.cpp]: Code exerpt taken from the openssl examples, What should be an standard nonce size?

File: src/libsync/clientsideencryption.cpp [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+
+void ClientSideEncryption::encryptPrivateKey(EVP_PKEY *keyPair)
+{
+    // Write the Private File to a BIO
+    // Retrieve the BIO contents, and encrypt it.
+    // Send the encrypted key to the server.
+    // I have no idea what I'm doing.
+
+    using ucharp = unsigned char *;
+    const char *salt = "$4$YmBjm3hk$Qb74D5IUYwghUmzsMqeNFx5z0/8$";
```
[the remark this answers is not in the document]
Comment (tcanabrava) [reply src/libsync/clientsideencryption.cpp]: // I have no idea what I'm doing. 
The salt used the same string as the android client.

File: src/libsync/wordlist.cpp [hunk 3]
```diff
@@ -0,0 +1,2082 @@
...
+        "youth",
+        "zebra",
+        "zero",
+        "zone",
+        "zoo"
+    };
+
+    QStringList randomWords;
+    while(randomWords.size() != nr) {
+        QString currWord = wordList.at(rand() % wordList.size());
```
[the remark this answers is not in the document]
Comment (tcanabrava) [reply src/libsync/wordlist.cpp]: I could use arc4random but then that's not avaliable in windows platforms. what should I use in windows?

[the remark this answers is not in the document]
Comment (treba123) [reply src/libsync/wordlist.cpp]: `rand_s()` is the cryptographically save variant on windows I think.
https://msdn.microsoft.com/en-us/library/sxtz2fa8.aspx

**Pull request closed without merging**